### PR TITLE
Qualcomm AI Engine Direct - Enable quantized VIT

### PR DESCRIPTION
**Pull request by shewu-quic**
Summary:
- Refactor and add the missing op builders
  - Add bmm, gelu, squeeze and layer_norm op builder and unit test
  - Refactor unsqueeze to support 4D -> 5D
- Refactor the passes
  - Support bmm pattern in convert_addmm_back_to_linear.py
  - Add the pass convert_binary_op_with_scalar.py to replace torch.ops.aten.xxx.Scalar with torch.ops.aten.xxx.Tensor
  Since torch.ops.aten.xxx.Scalar doesn't generate a placeholder node
for scalar after to_edge.
  - Add the pass convert_scaled_dot_product_attention.py to decompose from scaled_dot_product_attention to multiple nodes for quantized annotation.
  - Add the summary for reduce_dynamic_range.py
- Add the example script to run quantized torchvision vit

### CODE REVIEW

**Comment by kimishpatel** [comment]
Can you report the results?

**Comment by shewu-quic** [comment]
> Can you report the results?

The following results are accuracy and performance for quantized VIT in Qnn 2.16.
image
image

File: backends/qualcomm/passes/convert_addmm_back_to_linear.py [hunk 1]
```diff
@@ -17,6 +20,10 @@
 
 
 class ConvertAddmmmmWithLinear(ExportPass):
+    """
+    Handle missing quantization tag for addmm op after decomposing
```
Comment (kimishpatel) [inline backends/qualcomm/passes/convert_addmm_back_to_linear.py:24]: are you using quantization_tag for identifying nodes pattern? If so addmm should already have that tag.

File: backends/qualcomm/passes/convert_addmm_back_to_linear.py [hunk 2]
```diff
@@ -101,6 +115,26 @@ def _convert_addmm(self, gm: torch.fx.GraphModule, src_partition: SourcePartitio
             for user in addmm_node.users.copy():
                 user.replace_input_with(addmm_node, linear_node)
 
+    def _convert_bmm(self, gm: torch.fx.GraphModule, src_partition: SourcePartition):
+        output = src_partition.output_nodes[0]
+        bmm_node = [n for n in src_partition.nodes if n.target == self.bmm][0]
```
Comment (kimishpatel) [inline backends/qualcomm/passes/convert_addmm_back_to_linear.py:120]: There should only be one of such node, right? Dont you want to assert that?

File: backends/qualcomm/passes/convert_addmm_back_to_linear.py [hunk 3]
```diff
@@ -101,6 +115,26 @@ def _convert_addmm(self, gm: torch.fx.GraphModule, src_partition: SourcePartitio
...
+        add_node = [n for n in src_partition.nodes if n.target == self.add][0]
+
+        # weight -> expand_copy -> view_copy -> input of bmm
+        weight_node = bmm_node.args[1].args[0].args[0].args[0]
```
Comment (kimishpatel) [inline backends/qualcomm/passes/convert_addmm_back_to_linear.py:124]: How do you ensure that you actually do see this pattern?

Comment (kimishpatel) [inline backends/qualcomm/passes/convert_addmm_back_to_linear.py:124]: Also how do you distinguish if it is a weight node or really a bmm?

File: backends/qualcomm/passes/convert_scaled_dot_product_attention.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                decomposed_module = make_fx(
+                    node.target,
+                    decomposition_table=get_decompositions(
+                        [
+                            torch.ops.aten._scaled_dot_product_flash_attention.default,
+                        ]
+                    ),
+                    tracing_mode="fake",
+                    _allow_non_fake_inputs=True,
+                )(*input_tensors)
```
Comment (kimishpatel) [inline backends/qualcomm/passes/convert_scaled_dot_product_attention.py:36]: what about the other args?

File: backends/qualcomm/passes/convert_scaled_dot_product_attention.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+import torch
+from executorch.exir.pass_base import ExportPass, PassResult
+from torch._decomp import get_decompositions
+from torch.fx.experimental.proxy_tensor import make_fx
+
+
+class ConvertScaledDotProductAttention(ExportPass):
```
Comment (kimishpatel) [inline backends/qualcomm/passes/convert_scaled_dot_product_attention.py:12]: Why are you not doing this via graph rewrite? This kind of manual rewrite seems cumbersome. Also if this is for quantization purpose then we also have https://github.com/pytorch/pytorch/blob/main/test/quantization/pt2e/test_quantize_pt2e.py#L1255

File: backends/qualcomm/tests/models.py [hunk 6]
```diff
@@ -301,6 +342,17 @@ def forward(self, x):
         return x.reshape(1, 12)
 
 
+class ScaledDotProductAttention(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, query_layer, key_layer, value_layer):
+        attn_output = torch.nn.functional.scaled_dot_product_attention(
+            query_layer, key_layer, value_layer
```
Comment (kimishpatel) [inline backends/qualcomm/tests/models.py:351]: Are you not testing with varying other args including mask?

File: backends/qualcomm/tests/test_qnn_delegate.py [hunk 7]
```diff
@@ -231,6 +246,12 @@ def test_qnn_backend_mean_dim(self):
             with self.subTest(i=i):
                 self.lower_module_and_test_output(module, sample_input)
 
+    @unittest.skip("it will hang in runtime")
```
Comment (kimishpatel) [inline backends/qualcomm/tests/test_qnn_delegate.py:249]: Should you at least test quantization? And maybe separate quantization tests from runtime? I presume right now you cannot enable these in CI due to dependency on the device/libs

File: backends/qualcomm/tests/test_qnn_delegate.py [hunk 8]
```diff
@@ -231,6 +246,12 @@ def test_qnn_backend_mean_dim(self):
...
+    def test_qnn_backend_mha(self):
+        module = MultiheadAttention()  # noqa: F405
```
Comment (kimishpatel) [inline backends/qualcomm/tests/test_qnn_delegate.py:251]: is this meant to test both linear and sdpa layers in MHA?

File: backends/qualcomm/utils/utils.py [hunk 9]
```diff
@@ -54,11 +57,15 @@ def capture_program(
     module: torch.nn.Module,
     inputs: Tuple[torch.Tensor],
 ) -> exir.ExirExportedProgram:
-    ex_prog = exir.capture(
+    exir_exported_program = exir.capture(
```
Comment (kimishpatel) [inline backends/qualcomm/utils/utils.py:60]: So we should really be moving to export + to_edge api, like here https://github.com/pytorch/executorch/blob/main/backends/xnnpack/test/tester/tester.py#L146 or maybe i missed something. I dont know how you are doing quantization otherwise.

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Left some comments

**Comment by kimishpatel** [comment]
You might wanna rebase. I see some failures

File: backends/qualcomm/passes/convert_scaled_dot_product_attention.py (hunk 5 above)
Comment (shewu-quic) [reply backends/qualcomm/passes/convert_scaled_dot_product_attention.py:12]: I’m sorry, I don’t quite understand. Could you please explain what you mean by graph rewrite? Because I want to decompose sdpa to annotate quantization_annotation for each node in sdpa, I refer the code in pytorch.
https://github.com/pytorch/pytorch/blob/13473df0d770e67f92cb4bb1b728f3746f160b7c/test/test_decomp.py#L941

Actually we run this pass intransform_for_annotation.
https://github.com/CodeLinaro/executorch/blob/2dc87fe8352299e77701ab8e9019201f545a9a1b/backends/qualcomm/qnn_quantizer.py#L322

File: backends/qualcomm/tests/models.py (hunk 6 above)
Comment (shewu-quic) [reply backends/qualcomm/tests/models.py:351]: I just test the functionality of sdpa for now. Maybe we would add more test sets incoming.

File: backends/qualcomm/tests/test_qnn_delegate.py (hunk 7 above)
Comment (shewu-quic) [reply backends/qualcomm/tests/test_qnn_delegate.py:249]: I have test both of quantized and float-pointed operator.
https://github.com/CodeLinaro/executorch/blob/2dc87fe8352299e77701ab8e9019201f545a9a1b/backends/qualcomm/tests/test_qnn_delegate.py#L640

File: backends/qualcomm/tests/test_qnn_delegate.py (hunk 8 above)
Comment (shewu-quic) [reply backends/qualcomm/tests/test_qnn_delegate.py:251]: No, we just test the functionality of MHA.
https://github.com/CodeLinaro/executorch/blob/2dc87fe8352299e77701ab8e9019201f545a9a1b/backends/qualcomm/tests/models.py#L297

File: backends/qualcomm/passes/convert_addmm_back_to_linear.py (hunk 1 above)
Comment (shewu-quic) [reply backends/qualcomm/passes/convert_addmm_back_to_linear.py:24]: I hope I didn’t confuse you. The quantization tag here refers to the quantization annotation.
https://github.com/CodeLinaro/executorch/blob/2dc87fe8352299e77701ab8e9019201f545a9a1b/backends/qualcomm/qnn_quantizer_utils.py#L25C25-L25C48

File: backends/qualcomm/passes/convert_addmm_back_to_linear.py (hunk 3 above)
Comment (shewu-quic) [reply backends/qualcomm/passes/convert_addmm_back_to_linear.py:124]: Thank you for bringing this to our attention. The pattern you mentioned was found in VIT. However, we are still investigating how to check all patterns for linear. If you have any recommendations, I would be happy to hear them.

**Comment by shewu-quic** [comment]
> You might wanna rebase. I see some failures

Thanks for your reminder. I have rebased my branch.

**Comment by kimishpatel** [comment]
@shewu-quic is CI is running. Let me know when clean. WIll stamp.

**Comment by haowhsu-quic** [comment]
Thank you for reviewing. @shewu-quic is now on his PTO and will come back after 2/25.

**Comment by kimishpatel** [comment]
There is a lint error @haowhsu-quic

**Comment by kimishpatel** [comment]
@haowhsu-quic can you rebase? my apologies btw that this has taken so long

**Review by kimishpatel — APPROVED** [review]

**Comment by haowhsu-quic** [comment]
Hi @kimishpatel, thank you for reviewing. Since @shewu-quic is on his PTO, I create another PR(https://github.com/pytorch/executorch/pull/1839) and rebase it to the latest commit.
Could you also help merge https://github.com/pytorch/executorch/pull/1773? This is a blocker for us currently, thank you.

**Comment by kimishpatel** [comment]
On 1773 can you respond to lint error comment?

From: haowhsu-quic ***@***.***>
Date: Monday, February 5, 2024 at 8:09 AM
To: pytorch/executorch ***@***.***>
Cc: Kimish Patel ***@***.***>, Mention ***@***.***>
Subject: Re: [pytorch/executorch] Qualcomm AI Engine Direct - Enable quantized VIT (PR #1442)
Hi @kimishpatel, thank you for reviewing. Since @shewu-quic is on his PTO, I create another PR(#1839) and rebase it to the latest commit. Could you also help merge #1773? This is a blocker for us currently, thank you. — Reply to this
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender

ZjQcmQRYFpfptBannerEnd

Hi @kimishpatel<https://github.com/kimishpatel>, thank you for reviewing. Since @shewu-quic<https://github.com/shewu-quic> is on his PTO, I create another PR(#1839<https://github.com/pytorch/executorch/pull/1839>) and rebase it to the latest commit.
Could you also help merge #1773<https://github.com/pytorch/executorch/pull/1773>? This is a blocker for us currently, thank you.
